mcstatus/querier.py

```diff
@@ -128,6 +128,22 @@ def from_connection(cls, response: Connection):
         data = {}
         players = []
 
+        # If hostname is set to unicode, using other parameters of read_ascii() may be in the wrong order
```

Comment:
Sweet! Now we have tests that capture the bug and those tests are passing :partying_face: 
When tests are passing that opens the opportunity to refactor the code. I'm happy to work through how I came to this code over a call if you'd like to learn some refactoring tricks from me. Here's what I was able to come up with:
```python
import re
#...
    @classmethod
    def from_connection(cls, response: Connection):
        response.read(len("splitnum") + 1 + 1 + 1)
        data = {}
        players = []

        while True:
            key = response.read_ascii()
            if key == "hostname":
                name = re.search(b"(.*)\x00gametype", response.received).group(1)
                # Since the query protocol does not properly support unicode, the hostname is still not resolved
                # correctly; however, this will avoid other parameter parsing errors.
                data[key] = response.read(len(name)).decode("ISO-8859-1")
                response.read(1)  # ignore null byte
            elif len(key) == 0:
                response.read(1)
                break
            else:
                value = response.read_ascii()
                data[key] = value

        response.read(len("player_") + 1 + 1)

        while True:
            name = response.read_ascii()
            if len(name) == 0:
                break
            players.append(name)

        return cls(data, players)
```
The tests still pass, and the code is a bit shorter and easier to follow to me. What do you think? Are there any other refactorings that you have an idea for?

Reply:
so cool! I think your solution is better. I have modified the code to your method. You can review again, if there is no problem, I hope to apply this fix as soon as possible, as far as I know there are a lot of people in China who have headaches because of this bug.